Per user persistent storage

ccristescu-reol opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes
| BC breaks?    | No
| Deprecations? | No
| Fixed tickets | N/A

In a multi user environment where multiple projects are located on the same instance(s)
if two or more users are deploying a project on the same environment, the generated hosts files
in the system tmp folder are overwritten or worse, the deploy will fail due to file permissions.

Make the persistent storage folder configurable or use the system tmp  as default if not possible.

antonmedv commented on src/Host/Storage.php:
```diff
@@ -14,6 +14,73 @@
+		// use the system temporary folder and the current pid as default 
+		// persistent storage in case we can't use the configured value
+		// or we can't create the default location
+		$tmp = sys_get_temp_dir() . '/' . posix_getpid();
```
This extension is not available on Windows platforms.
http://php.net/manual/en/intro.posix.php

ccristescu-reol replied:
For non posix compatible i'm going to use getenv.
getenv('HOMEDRIVE') . getenv('HOMEPATH') are available on windows 8+ 

It's this enough?

antonmedv commented on src/Host/Storage.php:
```diff
@@ -14,6 +14,73 @@
+	 * @return string
+	 * @throws Exception
+	 */
+	private static function _getPersistentStorageLocation() {
```
Rename `_getPersistentStorageLocation` to `getPersistentStorageLocation` and move this method to end of class.

antonmedv commented on src/Host/Storage.php:
```diff
@@ -14,6 +14,73 @@
+		$tmp = sys_get_temp_dir() . '/' . posix_getpid();
+
+		// get the location for the deployer configuration
+		$deployerConfig = $config->has('deployer_config') ? $config->get('deployer_config') : null;
```
It's over this PR, I think such functionality should done in separate PR.

Anyway I don't see way it's needed. Please remove everything related to deployer_config.

antonmedv commented on src/Host/Storage.php:
```diff
@@ -14,6 +14,73 @@
+
+			// use the home dir of the current user 
+			// and the repository name
+			$userInfo = posix_getpwuid(posix_getuid());
```
Same here, what about windows?

ccristescu-reol replied:
For non posix compatible i'm going to use getenv.
getenv('HOMEDRIVE'), getenv('HOMEPATH') are available on windows 8+ 
getenv('HOME') for *nix

It's Windows 8+ enough?

antonmedv commented on src/Host/Storage.php:
```diff
@@ -14,6 +14,73 @@
+		} 
+		
+		// we now have the repository name
+		$repository = str_replace('/', '_', substr($configRepository, (strrpos($configRepository, ':') + 1)));
```
What is it for?

ccristescu-reol replied:
Long story short: i have five webapps on the same instances (my staging). If i deploy two apps at the same time (using two terminals for example) the second one i launch will overwrite the content of the .dep files so, the first deployer will start running tasks in the locations set by the second one and everything ends up in a mess.

The .dep files should be stored per webapp so, the easiest way was to use the repository to generate a separate folder name.

antonmedv replied:
I think it's easier to create rundom files instead.

ccristescu-reol replied:
Easier yes... but then you will need an internal task to clean up temporary files older than N days or something simmilar.
This way it's cleaner... i would say :)

antonmedv replied:
But tmp cleaned by system itself

ccristescu-reol commented:
Windows 8+ and non posix systems compatibility, added alternatives to the posix_* calls:
- getmypid() for posix_getpid
- getenv()for posix_getpwuid

Method _getPersistentStorageLocation renamed to getPersistentStorageLocation and moved at then end.

Dropped the option to have a configurable path for the persistent storage folder.

ccristescu-reol commented:
Removed everything and replaced it with randomly generate file names in system tmp.
